### ISSUE TITLE
MonetaryAmount serialization

### ISSUE THREAD

**Issue by OthmanTenich** [zalando/jackson-datatype-money#64]
Hello,

I have an issue during the serialization of a MonetaryAmount.

When I try to serialize an amount of 100.00 EUR for example, the corresponding JSON is the following :
`{"amount":"1E+2","currency":"EUR"}`

I would like to get `{"amount":"100.00","currency":"EUR"}` instead.

In fact, what I expect is an amount formatted with the currency digit places number. If the currency has 3 digit places (like the TND) I want to have `{"amount":"100.000","currency":"TND"}`.

Maybe I missed something in the library that's doing this.

Thank you in advance.

**Comment by jntakpe** [comment]
Same issue here

**Comment by whiskeysierra** [comment]
Based on your example I can see that you serialize amounts as strings. There are two ways to do that:

1. Enable `JsonGenerator.Feature.WRITE_NUMBERS_AS_STRINGS` or
2. Configure the `QuotedDecimalNumberValueSerializer`

In case you use 1. you could also enable `JsonGenerator.Feature.WRITE_BIGDECIMAL_AS_PLAIN`. That will at least give you:

```json
{"amount":"100","currency":"EUR"}
```

If you want to use the number of fraction digits as dictated by the currency we would need to change the library for this, as the `QuotedDecimalNumberValueSerializer` only uses `BigDecimal#toString` right now (it even ignores `WRITE_BIGDECIMAL_AS_PLAIN`).

The (implementation) problem that I see right now is that we don't have currency information anymore at the point where we serialize the amount/number.

**Comment by OthmanTenich** [comment]
Thank you for your prompt response.

If I understood the code, the serialization is done in the `serialize()` method of `MonetaryAmountSerializer`

Instead of the following code line :
`generator.writeObjectField(names.getAmount(), number);`

Is it possible to pass a `String` to Jackson there ?
We can then pass the formatted string of the amount.

I wrote a method that do this :
```
private static String serializeAmount(MonetaryAmount input) {
        BigDecimal amount = input.getNumber().numberValueExact(BigDecimal.class);
        CurrencyUnit currency = input.getCurrency();
        DecimalFormat decimalFormat = (DecimalFormat) 
        NumberFormat.getNumberInstance(Locale.ENGLISH);
        decimalFormat.applyPattern("###.##");
        decimalFormat.setMaximumFractionDigits(currency.getDefaultFractionDigits());
        decimalFormat.setMinimumFractionDigits(currency.getDefaultFractionDigits());
        return decimalFormat.format(amount).replace(",", "");
}
```

We could replace the previous line of code with the following one :
`generator.writeObjectField(names.getAmount(), serializeAmount(value));`

**Comment by whiskeysierra** [comment]
This call

> generator.writeObjectField(names.getAmount(), number);

will ultimately delegate to either `DecimalNumberValueSerializer` or `QuotedDecimalNumberValueSerializer`. 

> Is it possible to pass a String to Jackson there ?

Possible yes, but not desirable. The decision whether the amount is serialized as a decimal number (default) or as a string (optional) should not be made by the `MonetaryAmountSerializer`.

Your issue is actually two different issues:
1. Don't use scientific notation. This is accomplished by using `WRITE_BIGDECIMAL_AS_PLAIN`. We may need to add support for this when using the `QuotedDecimalNumberValueSerializer`.
2. Use correct number of fraction digits. This is useless for the `amount` property as this is meant to be used for calculations, not for display. Did you take a look at https://github.com/zalando/jackson-datatype-money#formatting?

**Comment by OthmanTenich** [comment]
The use of correct number of fraction digits is important in a banking context. If the amount does not contain the correct number of fraction, it is considered wrong.

But I agree with you that in terms of calculation it is useless.

I checked the formatting classes of the library and they seem to be used for the `formatted` filed in the Json as follows : `{"amount":"1E+2","currency":"EUR","formatted":"EUR100.00"}`

If you have any clue or help it would be appreciated. I am new to this library.

Thank you for your time.

**Comment by whiskeysierra** [comment]
> I checked the formatting classes of the library and they seem to be used for the formatted filed in the Json as follows : {"amount":"1E+2","currency":"EUR","formatted":"EUR100.00"}

That is correct. To me it sounds like you're trying to format it in a specific way.

**Comment by OthmanTenich** [comment]
Yes. But I would like to get something like `{"amount":"100.00","currency":"EUR"}`.

I need the `amount` field to be formatted, not to add a field to the Json. Why have you put something to format the amount as a String if the `amount` field is used for calculation purposes only ?

**Comment by whiskeysierra** [comment]
> I need the amount field to be formatted

Why? It wouldn't make a difference in calculations and that should be the only purpose of this field. In JSON `100`, `100.00` and `1E+2` are all equivalent.

**Comment by OthmanTenich** [comment]
In Json 100, 100.00 and 1E+2 are equivalent. But not in a banking context. I don't see the point of writing the amount as a string in the Json then. Numbers in Javascript are int, float, etc. not strings

**Comment by OthmanTenich** [comment]
Sorry if I'm not clear

**Comment by OthmanTenich** [comment]
I left the office. I will post some code tomorrow.

**Comment by OthmanTenich** [comment]
Hi,

I'm working on a banking application that uses Json in REST APIs. I need to write the amount in Json with the right format. When you work with money, 1E+2 EUR is not correct. Any money api validation will reject this. The correct format is 100.00 EUR. So I cannot send `{"amount":"1E+2","currency":"EUR"}`. This is not a correct format when you manipulate money amounts. I absolutely need a format like `{"amount":"100.00","currency":"EUR"}`.

I hope that I was enough clear for you to understand my need.

Thank you again.

**Comment by tkrop** [comment]
Interesting discussion.

My 2 cents: From a representational perspective I would expect the library to serialize to
`{"amount":100.00,"currency":"EUR"}`, as this exactly expressing the monetary amount in a kind of standardized JSON. I would expect the number of digits to be expanded with `0` as is usual for the minor units, and I would expect the library to allow for fractions of minor units to any length as this may be necessary for financial applications.

Internally the amount in my opinion must always be represented as unlimited precision number for serialization as well as for deserialization. A library should accept strings that are convertible to unlimited precision numbers. Under this rules, a feature representing the amount as string is unnecessary but harmless.

Explained this, in my opinion any library that fails to follow this rules by default is broken.

**Comment by whiskeysierra** [comment]
> When you work with money, 1E+2 EUR is not correct. Any money api validation will reject this.

But your money api validators are not working on JSON values, are they?

**Comment by OthmanTenich** [comment]
They are

**Comment by OthmanTenich** [comment]
They validate the JSON format of the amount

**Comment by OthmanTenich** [comment]
oups.

The APIs check that the JSON format is correct before deserializing it to a Java object

**Comment by whiskeysierra** [comment]
Do you have an extension for Jackson or are you using a different library for deserializing?

**Comment by OthmanTenich** [comment]
I am using yours for now. But I have the issue of serialization. I try to solve this before thinking to the deserialization.

**Comment by whiskeysierra** [comment]
> The APIs check that the JSON format is correct before deserializing it to a Java object

> I am using yours for now. But I have the issue of serialization. I try to solve this before thinking to the deserialization.

So you're planning to validate the JSON value before deserializing, but you are not doing it right now?

**Comment by whiskeysierra** [comment]
What I am majorly confused about is why the format of the JSON value matters. I mean you're obviously using this library and it deserializes into `Money`. As long as `100`, `1E+2` and `100.00` all deserialize into the very same `Money` value, then I don't see the problem here.

**Comment by OthmanTenich** [comment]
Right now I deserialize the JSON amount into a class that looks like this :
```
public class InstructedAmount {
      private String amount;
      private String currency;
}
```
This allows me to keep the JSON string format of the amount in Java and then perform controls on it.

**Comment by OthmanTenich** [comment]
The JSON format matters because I consume other APIs that will reject my request if the JSON amount format is not as expected.

**Comment by OthmanTenich** [comment]
I know that it does not matter in my internal model because I store a MonetaryAmount. But it is important to have the right format in the JSON too.

**Comment by whiskeysierra** [comment]
> The JSON format matters because I consume other APIs that will reject my request if the JSON amount format is not as expected.

Ok, that was the missing piece of information. You need to control the format for those APIs since you don't control them. Make sense then. I'll prepare a PR. It will be part of the 2.x version that is not released yet.

**Comment by OthmanTenich** [comment]
Ok. Thank you.

Could you include the digits formatting depending on the currency default digits number ?

**Comment by whiskeysierra** [comment]
Yes, I'll make the format based on `CurrencyUnit.getDefaultFractionDigits()`.

**Comment by whiskeysierra** [comment]
@OthmanTenich @jntakpe Take a look at #66

**Comment by whiskeysierra** [comment]
@OthmanTenich @jntakpe I released 1.0.0-RC1 to central.
